### PR TITLE
qwen3 testing and parameter loading

### PR DESCRIPTION
**Pull request by chapman20j**
Added logits tests for Qwen3 forward passes
Updated parameter loading in params.py
Updated the rotary positional embeddings.

### CODE REVIEW

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            past_key_values=past_key_values,
+            use_cache=True,
+            cache_position=cache_position,
+            position_embeddings=position_embeddings,
+        )
+        if higher_precision:
+            for k in out:
+                if hasattr(out[k], "to"):
+                    out[k] = out[k].to(torch.float32)
+        return out
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:93]: In a testing scenario do we ever expect to not use higher precision?
If not why don't we remove this variable and cast all torch Tensors to be float32?

ex:
```
out = {
    k: v.to(torch.float32) if isinstance(v, torch.Tensor) else v
    for k, v in args.items()
}
```

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+    def _nnx_forward_logits(self, cache, tokens, dtype=None):
+        segment_ids = 1 * (tokens != self.tokenizer.pad_token_id)
+        x = self.nnx_model.embedder.encode(tokens)
+        if dtype is not None:
+            x = x.astype(dtype)
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:99]: Can we just always pass dtype and remove the conditional (`if dtype is not None`) and remove default None for dtype argument??

ditto for l103-104

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+from huggingface_hub import snapshot_download
+from transformers import AutoTokenizer
+from transformers.cache_utils import Cache, DynamicCache
+from transformers.masking_utils import create_causal_mask, create_sliding_window_causal_mask
+from transformers.models.qwen3 import Qwen3Config, Qwen3ForCausalLM, Qwen3PreTrainedModel
+
+from bonsai.models.qwen3 import modeling, params
+from bonsai.models.qwen3.tests.run_model import tokenize
+
+RUN_ALL = True
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:20]: Can we remove this and run all tests by default? We want all logits to be correct within threhold anyways.

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+
+class TestModuleForwardPasses(absltest.TestCase):
+    def setUp(self):
+        super().setUp()
+        model_name: str = "Qwen/Qwen3-0.6B"
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+        ## models
+        self.torch_model = Qwen3ForCausalLM.from_pretrained(model_name, dtype="auto")
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:30]: Can we have centralized boilerplate of `self.torch_model.eval()` and remove all `tm.eval()` below?

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    def test_embedder(self):
+        nm = self.nnx_model.embedder
+        tm = self.torch_model.model.embed_tokens
+        tm.eval()
+
+        tx = torch.randint(0, self.token_dim, size=(self.batch_size, self.num_input_tokens))
+        jx = tx.cpu().detach().numpy()
+
+        jy, ty = nm.encode(jx), tm(tx)
+        torch.testing.assert_close(torch.tensor(jy), ty)
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:137]: could we explicitly state desired rtol values?
ditto all else where applicable

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+            num_kv_heads=self.bonsai_config.num_kv_heads,
+            head_dim=self.bonsai_config.head_dim,
+            dtype=jnp.float32,
+        )
+
+        nnx_logits = self._nnx_forward_logits(nnx_cache, tokens, jnp.float32)
+        torch_inputs = self._process_hf_tokens(query)
+        torch_logits = self.torch_model(**torch_inputs).logits
+
+        self._check_batched(torch_inputs["left_pads"], torch_logits, nnx_logits)
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:371]: nit: rename for clarity (ex: `_check_batched_input_logits`)

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        tm = tm.to(torch.float32)
+
+        nnx_cache = modeling.init_cache(
+            num_layers=self.bonsai_config.num_layers,
+            batch_size=self.batch_size,
+            cache_size=self.cache_size,
+            num_kv_heads=self.bonsai_config.num_kv_heads,
+            head_dim=self.bonsai_config.head_dim,
+            dtype=jnp.float32,
+        )
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:156]: This is used multiple times with only `batch_size` changed -- how about creating `_init_nnx_cache(self, batch_size: int)` as a separate function?

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+        jy, ty = nm.encode(jx), tm(tx)
+        torch.testing.assert_close(torch.tensor(jy), ty)
+
+    @unittest.skipIf(not RUN_ALL, "skipping submodule tests")
+    def test_decoder_layer(self):
+        nm = self.nnx_model.layers[0]
+        tm = self.torch_model.model.layers[0]
+        tm.eval()
+
+        jx = jax.random.normal(jax.random.key(0), shape=(32, 5, 1024))
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:145]: Prefer declaring shape from configurations for readability. Ditto all else

ex:
`jx = jax.random.normal(jax.random.key(0), shape=(self.batch_size, self.num_input_tokens, self.bonsai_config.hidden_size), dtype=self.test_dtype)`

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+            jx = jax.random.normal(jax.random.key(0), shape=(32, 5, 1024))
+            tx = torch.tensor(jx)
+            tm = tm.to(torch.float32)
+
+            jy = nm(jx, nnx_cache[i], jnp.ones((32, 5)), 0)
+            torch_inputs = self._setup_torch_attn(tx, higher_precision=True)
+            ty = tm(**torch_inputs)
+
+            torch.testing.assert_close(torch.tensor(jy), ty, atol=1e-3, rtol=1e-3)
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py:186]: If we want to propagate the previous layer's output as input to the next layer and test, we could have something like this:


```
        jx = jax.random.normal(jax.random.key(0), shape=(self.batch_size, self.num_input_tokens, self.bonsai_config.hidden_size), dtype=self.test_dtype)
        tx = torch.tensor(np.array(jx))

        for i, (nm, tm) in enumerate(zip(self.nnx_model.layers, self.torch_model.model.layers)):
            tm = tm.to(torch.float32)
            jy = nm(jx, nnx_cache[i], segment_ids, right_pads)
            torch_inputs = self._setup_torch_attn(tx)
            ty = tm(**torch_inputs)[0]
            torch.testing.assert_close(torch.tensor(jy), ty, atol=1e-3, rtol=1e-3)
            jx = jy  # Propagate output as input to the next layer
            tx = torch.tensor(np.array(jx))  # Update torch input
```

File: bonsai/models/qwen3/tests/test_outputs.py [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+                "input_embeds": input_embeddings,
+                "attention_mask": attention_mask,
+                "cache_position": cache_position,
+                "past_key_values": past_key_values,
+                "position_ids": position_ids,
+            }
+            # Create the masks
+            causal_mask_mapping = {
+                "full_attention": create_causal_mask(**mask_kwargs),
+            }
```
Comment (jenriver) [inline bonsai/models/qwen3/tests/test_outputs.py]: Can we just have kwargs be expressed explicitly

```
# Create the masks
causal_mask_mapping = {
    "full_attention": create_causal_mask(
        config=self.torch_model.config,
        input_embeds=input_embeddings,
        attention_mask=attention_mask,
        cache_position=cache_position,
        past_key_values=past_key_values,
        position_ids=position_ids,
    ),
 }
```

**Review by jenriver — APPROVED** [review]
Thanks for the thorough tests and the fix! :)
